File: scala/src/ch1/ex108.scala
object ex108 {
  val factor = 1000000.0

  def square (x: Double) : Double = x * x

  def cube (x: Double) : Double = x * x * x

  def improve (guess: Double, x: Double) : Double = (2 * guess + x / square(guess)) / 3

  def is_good_enough (guess: Double, x: Double) : Boolean = Math.abs (cube(guess) - x) < (guess / factor)

  def cbrt_iter (guess: Double, x: Double) : Double =
    if (is_good_enough(guess, x)) guess else cbrt_iter(improve(guess, x), x)

  def cbrt (x: Double) : Double = cbrt_iter(1.0, x)
}

object ex108App extends App {
  val delta = 0.0001
  assert (Math.abs (ex108.cube(ex108.cbrt(27)) - 27) < delta)
  assert (Math.abs (ex108.cube(ex108.cbrt(0.00027)) - 0.00027) < delta)
  println ("done testing ex108")
}

File: haskell/ch1/Ex108.hs
module Ex108 where

sq x = x * x

cube x = x * x * x

improve guess x = (x / sq(guess) + 2 * guess) / 3.0

is_good_enough guess x = abs (cube guess - x) < (guess / 100000.0)

cbrt_iter guess x = if (is_good_enough guess x) then guess else cbrt_iter (improve guess x) x

cbrt x = cbrt_iter 1.0 x
